src/interp/binary-reader-interp.cc

```diff
@@ -1394,6 +1490,90 @@ Result BinaryReaderInterp::OnTableInitExpr(Index segment_index,
+  Index catch_depth;
+  CHECK_RESULT(validator_.OnRethrow(loc, Var(depth)));
+  CHECK_RESULT(validator_.GetCatchDepth(depth, &catch_depth));
+  // The rethrow opcode takes an index into the exception stack rather than
+  // the number of catch nestings, so we subtract one here.
+  istream_.Emit(Opcode::Rethrow, catch_depth - 1);
+  return Result::Ok;
+}
+
+Result BinaryReaderInterp::OnTryExpr(Type sig_type) {
+  CHECK_RESULT(validator_.OnTry(loc, sig_type));
+  // Push a label that tracks mapping of exn -> catch
+  PushLabel(LabelKind::Try, Istream::kInvalidOffset, Istream::kInvalidOffset,
+            func_->handlers.size());
+  func_->handlers.push_back(HandlerDesc{HandlerKind::Catch,
+                                        istream_.end(),
+                                        Istream::kInvalidOffset,
+                                        {},
+                                        {Istream::kInvalidOffset}});
+  return Result::Ok;
+}
+
+Result BinaryReaderInterp::OnCatchExpr(Index tag_index) {
+  CHECK_RESULT(validator_.OnCatch(loc, Var(tag_index), false));
+  Label* label = TopLabel();
+  HandlerDesc& desc = func_->handlers[label->handler_desc_index];
+  desc.kind = HandlerKind::Catch;
+  // Jump to the end of the block at the end of the previous try or catch.
+  Istream::Offset offset = label->offset;
+  istream_.Emit(Opcode::Br);
+  if (offset == Istream::kInvalidOffset) {
+    depth_fixups_.Append(label_stack_.size() - 1, istream_.end());
+  }
+  istream_.Emit(offset);
+  desc.end_offset = istream_.end();
+  label->kind = LabelKind::Block;
+  desc.catches.push_back(CatchDesc{tag_index, istream_.end()});
+  return Result::Ok;
+}
+
+Result BinaryReaderInterp::OnCatchAllExpr() {
+  CHECK_RESULT(validator_.OnCatch(loc, Var(), true));
+  Label* label = TopLabel();
+  HandlerDesc& desc = func_->handlers[label->handler_desc_index];
+  desc.kind = HandlerKind::Catch;
+  Istream::Offset offset = label->offset;
+  istream_.Emit(Opcode::Br);
+  if (offset == Istream::kInvalidOffset) {
+    depth_fixups_.Append(label_stack_.size() - 1, istream_.end());
+  }
+  istream_.Emit(offset);
+  desc.end_offset = istream_.end();
+  label->kind = LabelKind::Block;
+  desc.catch_all_offset = istream_.end();
+  return Result::Ok;
+}
+
+Result BinaryReaderInterp::OnDelegateExpr(Index depth) {
+  CHECK_RESULT(validator_.OnDelegate(loc, Var(depth)));
+  Label* label = TopLabel();
+  HandlerDesc& desc = func_->handlers[label->handler_desc_index];
+  desc.kind = HandlerKind::Delegate;
+  Istream::Offset offset = label->offset;
+  istream_.Emit(Opcode::Br);
+  if (offset == Istream::kInvalidOffset) {
+    depth_fixups_.Append(label_stack_.size() - 1, istream_.end());
+  }
+  istream_.Emit(offset);
+  desc.end_offset = istream_.end();
+  Label* target_label = GetNearestTryLabel(depth + 1);
+  assert(target_label);
+  desc.delegate_offset = target_label->handler_desc_index;
+  FixupTopLabel();
+  PopLabel();
```

Comment:
So I think most of the checks that `OnEndExpr` are doing are specific to how `end` works, in that it can be paired with different instructions. Also the `Label` it uses is actually a different data structure from the validator. For the stack size check, that's also only for the case in which it's a function body block I think.

I have added an `assert` though that makes sure that the current label is a `LabelKind::Try` instead of possibly a `LabelKind::Block`, which makes this code do more sanity checking that it's only a `try-delegate`. Let me know if that's not enough.

Reply:
Aren't these lines necessary though? https://github.com/WebAssembly/wabt/blob/d61a78988001cdc6d364d91cfb9b2088b9ae2798/src/interp/binary-reader-interp.cc#L1024-L1027

These check 1. if the label is valid and 2. the type and number of the values on the stack and the label's type matches. `try`-`delegate` can act like a block, so it can return values.

Reply:
If I'm understanding correctly, I think these checks are already done by the `OnDelegate` call near L1587. That function's here: https://github.com/WebAssembly/wabt/blob/5d4955c0e819ce0d8943824f3d458894be628fd3/src/type-checker.cc#L572 and it checks the label (`CHECK_RESULT(GetLabel(depth + 1, &label));`) and does `PopAndCheckSignature` and so on.

Reply:
Oh I see.. Thanks.